iothub/device/src/InternalClient.cs

```diff
@@ -57,58 +50,40 @@ internal class InternalClient : IDisposable
 
         private int _diagnosticSamplingPercentage;
 
-        protected internal InternalClient(
+        internal InternalClient(
             IotHubConnectionCredentials iotHubConnectionCredentials,
-            IotHubClientOptions iotHubClientOptions,
-            IDeviceClientPipelineBuilder pipelineBuilder)
+            IotHubClientOptions iotHubClientOptions)
         {
-            Argument.AssertNotNull(iotHubClientOptions, nameof(iotHubClientOptions));
-
             if (Logging.IsEnabled)
-                Logging.Enter(this, iotHubClientOptions.TransportSettings, pipelineBuilder, nameof(InternalClient) + "_ctor");
+                Logging.Enter(this, iotHubClientOptions?.TransportSettings, nameof(InternalClient) + "_ctor");
 
-            IotHubConnectionCredentials = iotHubConnectionCredentials;
-            _clientOptions = iotHubClientOptions;
-
-            if (!iotHubClientOptions.ModelId.IsNullOrWhiteSpace()
-                && iotHubClientOptions.TransportSettings is IotHubClientHttpSettings)
+            // Make sure client options is initialized.
+            if (iotHubClientOptions == default)
             {
-                throw new InvalidOperationException("Plug and Play is not supported over the HTTP transport.");
+                iotHubClientOptions = new();
             }
 
-            var pipelineContext = new PipelineContext
+            IotHubConnectionCredentials = iotHubConnectionCredentials;
+            _clientOptions = iotHubClientOptions;
+
+            _pipelineContext = new PipelineContext
             {
-                IotHubConnectionCredentials = iotHubConnectionCredentials,
-                ProductInfo = iotHubClientOptions.ProductInfo,
-                IotHubClientTransportSettings = iotHubClientOptions.TransportSettings,
-                ModelId = iotHubClientOptions.ModelId,
+                IotHubConnectionCredentials = IotHubConnectionCredentials,
+                ProductInfo = _clientOptions.ProductInfo,
+                IotHubClientTransportSettings = _clientOptions.TransportSettings,
+                ModelId = _clientOptions.ModelId,
                 MethodCallback = OnMethodCalledAsync,
                 DesiredPropertyUpdateCallback = OnDesiredStatePatchReceived,
                 ConnectionStatusChangeHandler = OnConnectionStatusChanged,
-                ModuleEventCallback = OnModuleEventMessageReceivedAsync,
-                DeviceEventCallback = OnDeviceMessageReceivedAsync,
             };
 
-            pipelineBuilder ??= BuildPipeline();
```

Comment:
Could have abstract methods for deriving classes to add code for init, that the base calls, so it can init the pipeline as best it can, let a deriving class add on, and then call build (so derived classes don't have to know to do that).

Reply:
I wanted to go with this approach initially, but this being set up in the constructor makes our implementation options slightly limited. I'll try it again to see if I can find a way to make this work.

Reply:
done